# Project 3 -- Vinay Gannamaneni

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

- ***Link to AI Chat History***: Please share a link to your chat if you used AI (ex. ChatGPT Shared Links)
**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Average critic score by genre and total sales by console

In [10]:
import pandas as pd
video_games = pd.read_csv("/anvil/projects/tdm/data/grouping/vgchartz-2024.csv")
video_games.head()

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update
0,/games/boxart/full_6510540AmericaFrontccc.jpg,Grand Theft Auto V,PS3,Action,Rockstar Games,Rockstar North,9.4,20.32,6.37,0.99,9.85,3.12,2013-09-17,NaN
1,/games/boxart/full_5563178AmericaFrontccc.jpg,Grand Theft Auto V,PS4,Action,Rockstar Games,Rockstar North,9.7,19.39,6.06,0.60,9.71,3.02,2014-11-18,2018-01-03
2,/games/boxart/827563ccc.jpg,Grand Theft Auto: Vice City,PS2,Action,Rockstar Games,Rockstar North,9.6,16.15,8.41,0.47,5.49,1.78,2002-10-28,NaN
3,/games/boxart/full_9218923AmericaFrontccc.jpg,Grand Theft Auto V,X360,Action,Rockstar Games,Rockstar North,NaN,15.86,9.06,0.06,5.33,1.42,2013-09-17,NaN
4,/games/boxart/full_4990510AmericaFrontccc.jpg,Call of Duty: Black Ops 3,PS4,Shooter,Activision,Treyarch,8.1,15.09,6.18,0.41,6.05,2.44,2015-11-06,2018-01-14


In [11]:
video_games.shape

(64016, 14)

In [12]:
genre_scores = video_games.groupby("genre", as_index=False)["critic_score"].mean()
genre_scores

,genre,critic_score
0,Action,6.831008
1,Action-Adventure,7.678378
2,Adventure,7.341058
3,Board Game,5.300000
4,Education,7.750000
5,Fighting,7.392424
6,MMO,7.781818
7,Misc,6.884593
8,Music,7.992593
9,Party,5.968750


This shows the average critic score for each genre. Python automatically ignores NA values when calculating the mean.

Here we can see that Sandbox has the highest average critic score.

In [16]:
console_sales = video_games.groupby("console", as_index=False)["total_sales"].sum()
console_sales.sort_values("total_sales", ascending=False)

,console,total_sales
53,PS2,1027.76
73,X360,859.79
54,PS3,839.70
52,PS,546.25
55,PS4,539.92
...,...,...
65,TG16,0.00
77,XS,0.00
78,ZXS,0.00
79,iOS,0.00


This shows the total sales (in millions) for each console.

As displayed in the table, PS2 has the highest number of total sales.

The ``.groupby()`` function in Python allows us to split our dataset into groups based on the values in one or more columns. After grouping, we can apply functions like ``.mean()`` or ``.sum()`` to calculate statistics for each group. This is useful because it lets us compare different categories efficiently without having to filter the data manually for each category.

## Flagging high and low rated games and finding standout sports titles

In [17]:
video_games['high_rated'] = video_games['critic_score'] >= 7

This creates a True/False column. Rows with critic_score of 7 or more get True, others get False.

In [21]:
video_games['low_rated'] = video_games['critic_score'] < 7

This shows the same for games with critic_score strictly less than 7.

In [18]:
video_games['high_rated'].value_counts()

high_rated
False    59689
True      4327
Name: count, dtype: int64

In [22]:
video_games['low_rated'].value_counts()

low_rated
False    61665
True      2351
Name: count, dtype: int64

In [24]:
sports_hits = video_games[(video_games['genre'] == 'Sports') & 
                          (video_games['high_rated'] == True)]
sports_hits[sports_hits['critic_score'] == 9.5]

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
75,/games/boxart/1473680ccc.jpg,Madden NFL 2004,PS2,Sports,EA Sports,EA Tiburon,9.5,5.23,4.26,0.01,0.26,0.71,2003-08-12,NaN,True,False
114,/games/boxart/4564394ccc.jpg,Madden NFL 2005,PS2,Sports,EA Sports,EA Tiburon,9.5,4.53,4.18,0.01,0.26,0.08,2004-08-09,NaN,True,False
4359,/games/boxart/8490939ccc.jpg,SSX 3,GC,Sports,EA Sports BIG,EA Canada,9.5,0.38,0.29,NaN,0.08,0.01,2003-10-20,NaN,True,False


This shows all Sports games that received exactly 9.5.

In [25]:
sports_fails = video_games[(video_games['genre'] == 'Sports') & 
                           (video_games['low_rated'] == True)]
sports_fails[sports_fails['critic_score'] == 2.8]

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
8292,/games/boxart/full_554728AmericaFrontccc.jpg,Major League Baseball 2K10,DS,Sports,2K Sports,Visual Concepts,2.8,0.15,0.14,NaN,NaN,0.01,2010-03-02,NaN,False,True
59297,/games/boxart/full_mlb-slam_460AmericaFront.jpg,MLB Slam!,NGage,Sports,THQ,THQ,2.8,NaN,NaN,NaN,NaN,NaN,2008-10-31,NaN,False,True


This shows all Sports games that received exactly 2.8.

## Counting games by genre and by console-genre combination

In [26]:
genre_counts = video_games.groupby("genre", as_index=False)["title"].count()

This counts how many games belong to each genre.

In [30]:
genre_counts

,genre,title
0,Action,8557
1,Action-Adventure,1877
2,Adventure,6260
3,Board Game,33
4,Education,35
5,Fighting,2367
6,MMO,115
7,Misc,9304
8,Music,297
9,Party,151


In [32]:
genre_counts.sort_values("title", ascending=False).head(10)

,genre,title
7,Misc,9304
0,Action,8557
2,Adventure,6260
13,Role-Playing,5721
17,Sports,5586
15,Shooter,5410
10,Platform,4001
18,Strategy,3685
11,Puzzle,3521
12,Racing,3425


Mischas the most games listed in the database with a count of 9304

In [34]:
console_genre_counts = video_games.groupby(["console", "genre"], as_index=False).size()
console_genre_counts.head()

,console,genre,size
0,2600,Action,302
1,2600,Adventure,3
2,2600,Fighting,3
3,2600,Misc,21
4,2600,Platform,17


This shows how many games exist for each console-genre combination.

## Cleaning malformed review counts in the Google Play Store data

In [36]:
app_store = pd.read_csv("/anvil/projects/tdm/data/grouping/googleplaystore.csv")
app_store.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [37]:
app_store.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

We can see that 'Reviews' is 'object' (string) instead of numeric.

In [41]:
app_store["Clean_Reviews"] = app_store["Reviews"].str.replace(",", "", regex=False)
app_store["Numeric_Reviews"] = app_store["Clean_Reviews"].astype(float)

ValueError: could not convert string to float: '3.0M'

The error message at the bottom shows "3.0M" is causing the problem. This value represents 3.0 million but cannot be converted to float.

In [43]:
app_store = app_store[app_store["Clean_Reviews"] != "3.0M"]
app_store["Numeric_Reviews"] = app_store["Clean_Reviews"].astype(float)

We filter out rows where Clean_Reviews equals "3.0M".

In [45]:
app_store["rating_per_review"] = app_store["Rating"] / app_store["Numeric_Reviews"]
app_store["rating_per_review"].head()

0    0.025786
1    0.004033
2    0.000054
3    0.000021
4    0.004447
Name: rating_per_review, dtype: float64

This creates a new column showing rating divided by number of reviews.

## Standardizing content rating labels and counting apps by rating

In [46]:
app_store["Content Rating"].unique()
app_store["Content Rating"].unique().shape

(6,)

There are 6 unique content rating values.

In [47]:
app_store["New_Content_Ratings"] = app_store["Content Rating"].str.upper()
app_store["New_Content_Ratings"].head()

0    EVERYONE
1    EVERYONE
2    EVERYONE
3        TEEN
4    EVERYONE
Name: New_Content_Ratings, dtype: object

All content ratings are now in uppercase letters.

In [48]:
app_store["New_Content_Ratings"].unique()

array(['EVERYONE', 'TEEN', 'EVERYONE 10+', 'MATURE 17+',
       'ADULTS ONLY 18+', 'UNRATED'], dtype=object)

In [49]:
app_store.groupby("New_Content_Ratings", as_index=False)["App"].count()

,New_Content_Ratings,App
0,ADULTS ONLY 18+,3
1,EVERYONE,8714
2,EVERYONE 10+,414
3,MATURE 17+,499
4,TEEN,1208
5,UNRATED,2


This shows how many apps fall into each content rating category. The category 'EVERYONE' has the most apps.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
